Fix #9945: Check if dynamic property is set-only when dotting off it

khyperia opened the pull request:
When a set-only property (whose type is dynamic) is accessed via another member access, CS0154 ("property cannot be used in this context because it lacks a get accessor") was not produced, and instead crashed during emit.

Ping @jaredpar @gafter for review.

jaredpar commented:
@dotnet/roslyn-compiler

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1685,8 +1685,21 @@ internal bool CheckValueKind(BoundExpression expr, BindValueKind valueKind, Diag
+                        // dynamic access needs to check if the source of the
+                        // dynamic object is from, for example, a set-only property
+                        var access = (BoundDynamicMemberAccess)expr;
+                        return CheckValueKind(access.Receiver, BindValueKind.RValue, diagnostics);
```
It feels like this check is too late. I would expect it to be performed before BoundDynamicMemberAccess is created. Here we are supposed to check BoundDynamicMemberAccess itself

VSadov replied:
Yes. It seems the Value check on the receiver should happen in BindDynamicMemberAccess.
CheckValueKind for BoundDynamicMemberAccess should just return true - since we do not know until runtime.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1685,8 +1685,21 @@ internal bool CheckValueKind(BoundExpression expr, BindValueKind valueKind, Diag
+                        var access = (BoundDynamicIndexerAccess)expr;
+                        if (access.ReceiverOpt != null)
+                        {
+                            return CheckValueKind(access.ReceiverOpt, BindValueKind.RValue, diagnostics);
```
Same here

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/DynamicTests.cs:
```diff
@@ -3434,6 +3434,33 @@ static void Main()
         }
 
+        [Fact, WorkItem(9945, "https://github.com/dotnet/roslyn/issues/9945")]
+        public void DynamicGetOnlyProperty()
```
Two places in binding were modified, but it looks like this test covers only one of them.

khyperia commented:
Updated from feedback. Aleksey's test comment is no longer relevant, only one line is now changed. Other comments addressed by moving the check to BindMemberAccessWithBoundLeft, which is where the same situation in the non-dynamic case is detected (very bottom of the method, only other place where CheckValue is called in that method).

Please re-review, thanks!

jmarolf commented:
retest vsi please

davkean commented:
retest roslyn_prtest_win_vsi0  please
Failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_vsi0/63/, looks like https://github.com/dotnet/roslyn/issues/10110.

davkean commented:
retest vsi please

gafter commented:
:+1:

VSadov commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/DynamicTests.cs:
```diff
@@ -3434,6 +3434,33 @@ static void Main()
+    static void Main()
+    {
+        I i = null;
+        System.Type t = i.d.GetType();
```
Please add a test for DynamicIndexerAccess
